flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/TransactionalIdsGenerator.java

```diff
@@ -55,7 +55,7 @@ public TransactionalIdsGenerator(
 		checkArgument(safeScaleDownFactor > 0);
 		checkArgument(subtaskIndex >= 0);
 
-		this.prefix = checkNotNull(prefix);
+		this.prefix = checkNotNull(prefix).replaceAll("%", "%%");
```

Comment:
Would it be better to keep the original String as is, but only replace `%` with `%%` when `String.format()` is called? This will ensure all the variables have a consistent view of a String.

Reply:
@becketqin sorry for lately reply, do you mean we should keep _this.prefix = checkNotNull(prefix);_ unchanged. but user use String.format to replace % with %% of the prefix before calling _TransactionalIdsGenerator_?

Reply:
@leesf I meant we could change `String.format(prefix + "-%d", transactionalId);` to `String.format(prefix.replaceAll("%", "%%") + “-%d”, transactionalId;` Since the only reason we need to replace % with %% is this case, then it might be better to limit the impact to this case. Otherwise, it may sometimes cause confusion. For example, if users take a heap dump, they may see the value of the variable `prefix` different from the value they have in the config.

If we want to avoid replacing a string every time when creating a new transactional ID (this is probably not a big issue), maybe a separate variable like `prefixForStringFormat` would be clearer.

Reply:
@becketqin Updated this PR to address your comments. Please take another look.